bootstrap/src/lib/cache/index.ts

```diff
@@ -70,6 +81,112 @@ export const redactOptions = (options: CacheOptions) => ({
       : local.redactOptions(options.cacheOptions),
 })
 
+type RateLimitParticipant = {
+  // for this this type of req, how many underlying requests to provider
+  cost: number
+  // importance of this type of req
+  weight: number
+}
+
+type RateLimitGroup = {
+  totalCapacity: number
+  participants: { [key: string]: RateLimitParticipant }
+}
+
+type RateLimitOptions = {
+  groupMaxAge: number
+  groupId: string | undefined
+  participantId: string
+  totalCapacity: number
+}
+
+const makeRateLimit = (
+  options: RateLimitOptions,
+  cache: local.LocalLRUCache | redis.RedisCache,
+) => {
+  const _isEnabled = () => {
+    return options.groupId && options.totalCapacity
+  }
+
+  const _getRateLimitGroup = async (): Promise<RateLimitGroup | undefined> => {
+    if (!options.groupId) return
+    const result: any = (await cache.get(options.groupId)) || {}
+    const _getMinimumCapacity = (current: number, upcoming: number) => {
+      if (!current) return upcoming
+      return Math.min(current, upcoming)
+    }
+    return {
+      totalCapacity: _getMinimumCapacity(result.totalCapacity, options.totalCapacity),
+      participants: result.participants || {},
+    }
+  }
+
+  const _updateRateLimitGroup = async (
+    cost = DEFAULT_RATE_COST,
+    weight = DEFAULT_RATE_WEIGHT,
+  ): Promise<RateLimitGroup | undefined> => {
+    if (!options.groupId) return
+    const rateLimitGroup = await _getRateLimitGroup()
+    if (!rateLimitGroup) return
+    const newGroup = {
+      totalCapacity: rateLimitGroup.totalCapacity,
+      participants: {
+        ...rateLimitGroup.participants,
+        [options.participantId]: {
+          cost,
+          weight,
+        },
+      },
+    }
+    if (cache instanceof local.LocalLRUCache) {
+      await cache.set(options.groupId, newGroup, options.groupMaxAge)
+    } else {
+      await cache.setKeepingMaxAge(options.groupId, newGroup, options.groupMaxAge)
```

Comment:
Keeping the same age as it was before updating the group avoids renewing an old participant when a new one joins the group